app/controllers/api/api_controller.rb

```diff
@@ -40,17 +41,51 @@ class ApiController < ApplicationController
       RestPack::Serializer::InvalidInclude,
       ActiveRecord::RecordNotUnique, with: :unprocessable_entity
 
+    prepend_before_action :authenticate!
     prepend_before_action :require_login, only: [:create, :update, :destroy]
     prepend_before_action :ban_user, only: [:create, :update, :destroy]
     prepend_before_action ContentTypeFilter.new(*API_ACCEPTED_CONTENT_TYPES,
                                                 API_ALLOWED_METHOD_OVERRIDES)
 
-    def current_resource_owner
-      if doorkeeper_token
-        @current_resource_owner ||= User.find_by_id(doorkeeper_token.resource_owner_id)
+    def self.require_authentication(*actions, scopes: [])
+      @doorkeeper_scopes_per_action ||= {}
+      @doorkeeper_scopes_per_action[actions.map(&:to_s)] = scopes
+    end
+
+    def self.doorkeeper_scopes_per_action
+      @doorkeeper_scopes_per_action
+    end
+
+    def authenticate!
+      case
+      when has_basic_credentials?(request)
+        authenticate_or_request_with_http_basic do |login, password|
+          if user = User.admin_login_using_basic_auth(login, password)
+            @current_resource_owner = user
+          else
+            raise Api::NotLoggedIn
+          end
+        end
+      when doorkeeper_token
+        check_doorkeeper_authentication!
+        @current_resource_owner = User.find_by_id(doorkeeper_token.resource_owner_id)
+      else
+        check_doorkeeper_authentication!
       end
     end
 
+    def check_doorkeeper_authentication!
+      (self.class.doorkeeper_scopes_per_action || {}).each do |actions, scopes|
+        if actions.empty? || actions.include?("all") || actions.include?(params[:action])
```

Comment:
Talked this through IRL, we're going to make it so that you specify both public and private actions. Unspecified actions raise an exception and are thereby completely inaccessible. Maybe even decorator-style:

```ruby
auth :public \
def index; end

auth :private, scopes: [:project] \
def show; end
```